packages/material-ui/src/Menu/Menu.js

```diff
@@ -71,13 +101,44 @@ const Menu = React.forwardRef(function Menu(props, ref) {
     }
   };
 
+  const handleEntered = (element, isAppearing) => {
+    if (atLeastOneSubMenu) setEntering(false);
+
+    if (onEntered) {
+      onEntered(element, isAppearing);
+    }
+  };
+
+  const handleOnClose = (event) => {
+    event.preventDefault();
+    setOpenSubMenuIndex(null);
+    if (onClose) {
+      onClose(event, `${event.key.toLowerCase()}KeyDown`);
+    }
+  };
+
   const handleListKeyDown = (event) => {
-    if (event.key === 'Tab') {
-      event.preventDefault();
+    const { anchorEl } = other;
+    if (event.key === 'Tab' || event.key === 'Escape') {
+      handleOnClose(event);
+    }
 
-      if (onClose) {
-        onClose(event, 'tabKeyDown');
-      }
+    // This removes the imperatively added focusVisible class (from just below)
+    // when the user navigates away from the item and allows normal focus
+    // management to take back over.
+    anchorEl.onblur = () => anchorEl.classList.remove('Mui-focusVisible');
+
+    if (event.key === 'ArrowLeft' && isSubMenu) {
+      // This assigns the focusVisible class to the parent item when closing
+      // a sub menu using the left arrow keyboard navigation.
+      anchorEl.onfocus = async () => {
+        await anchorEl.classList.remove('MuiMenuItem-openSubMenuParent');
+        anchorEl.classList.add('Mui-focusVisible');
+      };
```

Comment:
Why do we need to mutate the DOM here? This makes it incredibly hard to reason about the component if an element is controlled by React and 3rd party code.

Reply:
This was recently added to fix the regression caused by the changes discussed in [this comment](https://github.com/mui-org/material-ui/pull/20591#issuecomment-718926526) that had this PR hung up for a long while. We tried a number of different ways to apply the focused style onto the parent item when we arrow left from a submenu, but couldn't get it to work without some imperatives... I'd love to find a better way. We looked at this for a few weeks, tried a number of different approaches and this was the only thing that worked that we tried.

Reply:
Could you point me to a test for that specific behavior. Or some JSX + steps to reproduce that I can try in a codesandbox.

Reply:
Apologies, I'm realizing I missed several of your responses that were buried in collapsed parts of the conversation and I haven't been getting emails for all of these, it seems.

The exact assertion that breaks without this is https://github.com/EsoterikStare/material-ui/blob/master/packages/material-ui/src/Menu/Menu.test.js#L956